server/src/middleware/mod.rs

```diff
@@ -33,6 +33,211 @@ mod host_filter;
 /// Proxy `GET /path` to internal RPC methods.
 mod proxy_get_request;
 
+use std::{net::SocketAddr, sync::Arc};
+
 pub use authority::*;
 pub use host_filter::*;
 pub use proxy_get_request::*;
+
+use http::{HeaderMap, Uri};
+use tower::layer::util::{Identity, Stack};
+use tower::layer::LayerFn;
+use tower::util::Either;
+
+use jsonrpsee_core::server::{
+	BoundedSubscriptions, MethodCallback, MethodResponse, MethodSink, Methods, SubscriptionState,
+};
+use jsonrpsee_core::traits::IdProvider;
+use jsonrpsee_types::error::{reject_too_many_subscriptions, ErrorCode};
+use jsonrpsee_types::{ErrorObject, Request};
+
+/// The transport protocol used to send or receive a call or request.
+#[derive(Debug, Copy, Clone)]
+pub enum TransportProtocol {
+	/// HTTP transport.
+	Http,
+	/// WebSocket transport.
+	WebSocket,
+}
+
+impl std::fmt::Display for TransportProtocol {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		let s = match self {
+			Self::Http => "http",
+			Self::WebSocket => "websocket",
+		};
+
+		write!(f, "{s}")
+	}
+}
+
+/// Context to a JSON-RPC call.
+#[derive(Debug, Clone)]
+pub struct Context {
+	/// Transport protocol.
+	pub transport: TransportProtocol,
+	/// Remote addr.
+	pub remote_addr: SocketAddr,
+	/// Connection id.
+	pub conn_id: usize,
+	/// HTTP Headers.
+	pub headers: HeaderMap,
+	/// URI.
+	pub uri: Uri,
+}
+
+/// Configuration of the RpcService.
+#[derive(Clone, Debug)]
+pub(crate) enum RpcServiceCfg {
+	/// The server supports only calls.
+	OnlyCalls,
+	/// The server supports both method calls and subscriptions.
+	CallsAndSubscriptions {
+		bounded_subscriptions: BoundedSubscriptions,
+		sink: MethodSink,
+		id_provider: Arc<dyn IdProvider>,
+		_pending_calls: tokio::sync::mpsc::Sender<()>,
+	},
+}
+
+/// JSON-RPC service middleware.
+#[derive(Clone, Debug)]
+pub struct RpcService {
+	conn_id: usize,
+	methods: Methods,
+	max_response_body_size: usize,
+	cfg: RpcServiceCfg,
+}
+
+impl RpcService {
+	/// Create a new service with doesn't support subscriptions.
+	pub(crate) fn new(methods: Methods, max_response_body_size: usize, conn_id: usize, cfg: RpcServiceCfg) -> Self {
+		Self { methods, max_response_body_size, conn_id, cfg }
+	}
+}
+
+/// Similar to `tower::Service` but specific for jsonrpsee and
+/// doesn't requires `&mut self` for performance reasons.
+#[async_trait::async_trait]
+pub trait RpcServiceT<'a> {
+	/// Process a single JSON-RPC call it may be a subscription or regular call.
+	/// In this interface they are treated in the same way but it's possible to
+	/// distinguish those based on the `MethodResponse`.
+	async fn call(&self, request: Request<'a>, ctx: &Context) -> MethodResponse;
```

Comment:
another comment from James is that this is leaky to expose `Context` here

would be nice to just do

```rust
let m = RpcServiceBuilder::new().layer_fn(|service, ctx| MyMiddleware { service, ctx });
```
but I had a look and I didn't manage to get it working....

Reply:
Ah, what issue did you run into?

I don't really think it's a big deal either way, but just felt cleaner passing the connection details on creation like that because semantically they will never change, and then the middleware can either take a copy of whatever it cares about from them or just ignore them entirely :)

(also I guess it offers the possibility to make a note as soon as a new connection is established or whatever, but that's less of a deal)